azure/services/natgateways/natgateways.go

```diff
@@ -0,0 +1,148 @@
+
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-azure/azure"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/scope"
+	"sigs.k8s.io/cluster-api-provider-azure/util/tele"
+)
+
+// NatGatewayScope defines the scope interface for nat gateway service
+type NatGatewayScope interface {
+	logr.Logger
+	azure.ClusterDescriber
+	azure.NetworkDescriber
+	NatGatewaySpecs() []azure.NatGatewaySpec
+}
+
+// Service provides operations on azure resources
+type Service struct {
+	Scope NatGatewayScope
+	client
+}
+
+// New creates a new service.
+func New(scope *scope.ClusterScope) *Service {
+	return &Service{
+		Scope:  scope,
+		client: newClient(scope),
+	}
+}
+
+// Reconcile gets/creates/updates a nat gateway.
+func (s *Service) Reconcile(ctx context.Context) error {
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Reconcile")
+	defer span.End()
+
+	if !s.Scope.Vnet().IsManaged(s.Scope.ClusterName()) {
+		s.Scope.V(4).Info("Skipping nat gateways reconcile in custom vnet mode")
+		return nil
+	}
+
+	for _, natGatewaySpec := range s.Scope.NatGatewaySpecs() {
+		existingNatGateway, err := s.getExisting(ctx, natGatewaySpec)
+
+		switch {
+		case err != nil && !azure.ResourceNotFound(err):
+			return errors.Wrapf(err, "failed to get nat gateway %s in %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+		case err == nil:
+			// nat gateway already exists, leave it alone
+			if natGatewaySpec.Subnet.Role == infrav1.SubnetControlPlane {
+				existingNatGateway.DeepCopyInto(s.Scope.ControlPlaneNatGateway())
+			} else if natGatewaySpec.Subnet.Role == infrav1.SubnetNode {
+				existingNatGateway.DeepCopyInto(s.Scope.NodeNatGateway())
+			} else {
+				continue
+			}
+		default:
+			// nat gateway doesn't exist but it was specified in the subnet, let's create it
+			s.Scope.V(2).Info("creating Nat Gateway", "nat gateway", natGatewaySpec.Name)
+			err = s.client.CreateOrUpdate(
+				ctx,
+				s.Scope.ResourceGroup(),
+				natGatewaySpec.Name,
+				network.NatGateway{
+					Location: to.StringPtr(s.Scope.Location()),
+					NatGatewayPropertiesFormat: &network.NatGatewayPropertiesFormat{
+						PublicIPAddresses: &[]network.SubResource{
+							{
+								ID: to.StringPtr(natGatewaySpec.NatGatewayIP.Name),
+							},
+						},
+					},
+				},
+			)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create nat gateway %s in resource group %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+			}
+			s.Scope.V(2).Info("successfully created nat gateway", "nat gateway", natGatewaySpec.Name)
+		}
+	}
+	return nil
+}
+
+func (s *Service) getExisting(ctx context.Context, spec azure.NatGatewaySpec) (*infrav1.NatGateway, error) {
+	existingNatGateway, err := s.Get(ctx, s.Scope.ResourceGroup(), spec.Name)
+	if err != nil {
+		if azure.ResourceNotFound(err) {
+			return nil, err
+		}
+		return nil, errors.Wrapf(err, "failed to get Nat Gateway %s", spec.Name)
+	}
+
+	return &infrav1.NatGateway{
```

Comment:
Can we move this logic to converters?

Reply:
I've seen converters used for common things like tags or SKU, but most scopes don't use them?